File: ruby/spec/phrase_spec.rb
# frozen_string_literal: true

require 'asciidoctor'
require_relative '../lib/curs/asciidoctor/extensions/phrase_inline_macro'

describe Curs::AsciidoctorExtensions::PhraseInlineMacro do
  context 'convert to html5' do
    it 'should apply mark formatting on inline phrases' do
      input = <<~'ADOC'
        phrase:mark[text,role="fragment highlight-red"]
      ADOC
      output = Asciidoctor.convert(input, doctype: 'inline')
      expect(output).to eql '<mark class="fragment highlight-red">text</mark>'
    end
    it 'should apply monospaced formatting on inline phrases' do
      input = <<~'ADOC'
        phrase:monospaced[text,role="fragment highlight-red"]
      ADOC
      output = Asciidoctor.convert(input, doctype: 'inline')
      expect(output).to eql '<code class="fragment highlight-red">text</code>'
    end
    it 'should apply emphasis formatting on inline phrases' do
      input = <<~'ADOC'
        phrase:emphasis[text,role="fragment highlight-red"]
      ADOC
      output = Asciidoctor.convert(input, doctype: 'inline')
      expect(output).to eql '<em class="fragment highlight-red">text</em>'
    end
    it 'should apply strong formatting on inline phrases' do
      input = <<~'ADOC'
        phrase:strong[text,role="fragment highlight-red"]
      ADOC
      output = Asciidoctor.convert(input, doctype: 'inline')
      expect(output).to eql '<strong class="fragment highlight-red">text</strong>'
    end
    it 'should apply superscript formatting on inline phrases' do
      input = <<~'ADOC'
        phrase:superscript[text,role="fragment highlight-red"]
      ADOC
      output = Asciidoctor.convert(input, doctype: 'inline')
      expect(output).to eql '<sup class="fragment highlight-red">text</sup>'
    end
    it 'should apply subscript formatting on inline phrases' do
      input = <<~'ADOC'
        phrase:subscript[text,role="fragment highlight-red"]
      ADOC
      output = Asciidoctor.convert(input, doctype: 'inline')
      expect(output).to eql '<sub class="fragment highlight-red">text</sub>'
    end
    it 'should apply formatting on inline phrases' do
      input = <<~'ADOC'
        phrase:#[text,role="fragment highlight-red"]
      ADOC
      output = Asciidoctor.convert(input, doctype: 'inline')
      expect(output).to eql '<span class="fragment highlight-red">text</span>'
    end
  end
  context 'load to document' do
    it 'should preserve attributes' do
      input = <<~'ADOC'
        phrase:strong[text,data-fragment-index=1]
      ADOC

      class CustomHtml5Converter < (Asciidoctor::Converter.for 'html5')
        register_for 'html5'

        def convert_inline_quoted(node)
          data_attrs = (node.attributes.select do |key, _|
            key.to_s.start_with? 'data-'
          end).map do |key, value|
            %(#{key}="#{value}")
          end.join(' ')
          %(<#{node.type}#{data_attrs ? " #{data_attrs}" : ''}>#{node.text}</#{node.type}>)
        end
      end

      doc = Asciidoctor.load(input, doctype: 'inline')
      expect(doc.convert).to eql '<strong data-fragment-index="1">text</strong>'
    end
  end
end

File: spec/phrase.spec.js
/* global describe it */
const chai = require('chai')
const expect = chai.expect

const phraseInlineMacroExtension = require('../lib/phrase-inline-macro.js')
const asciidoctor = require('@asciidoctor/core')()

describe('Phrase inline macro extension', () => {
  it('should apply mark formatting on inline phrases', () => {
    const registry = asciidoctor.Extensions.create()
    phraseInlineMacroExtension.register(registry)
    const input = 'phrase:mark[text,role="fragment highlight-red"]'
    const output = asciidoctor.convert(input, { doctype: 'inline', extension_registry: registry })
    expect(output).to.equal('<mark class="fragment highlight-red">text</mark>')
  })
  it('should apply monospaced formatting on inline phrases', () => {
    const registry = asciidoctor.Extensions.create()
    phraseInlineMacroExtension.register(registry)
    const input = 'phrase:monospaced[text,role="fragment highlight-red"]'
    const output = asciidoctor.convert(input, { doctype: 'inline', extension_registry: registry })
    expect(output).to.equal('<code class="fragment highlight-red">text</code>')
  })
  it('should apply emphasis formatting on inline phrases', () => {
    const registry = asciidoctor.Extensions.create()
    phraseInlineMacroExtension.register(registry)
    const input = 'phrase:emphasis[text,role="fragment highlight-red"]'
    const output = asciidoctor.convert(input, { doctype: 'inline', extension_registry: registry })
    expect(output).to.equal('<em class="fragment highlight-red">text</em>')
  })
  it('should apply strong formatting on inline phrases', () => {
    const registry = asciidoctor.Extensions.create()
    phraseInlineMacroExtension.register(registry)
    const input = 'phrase:strong[text,role="fragment highlight-red"]'
    const output = asciidoctor.convert(input, { doctype: 'inline', extension_registry: registry })
    expect(output).to.equal('<strong class="fragment highlight-red">text</strong>')
  })
  it('should apply superscript formatting on inline phrases', () => {
    const registry = asciidoctor.Extensions.create()
    phraseInlineMacroExtension.register(registry)
    const input = 'phrase:superscript[text,role="fragment highlight-red"]'
    const output = asciidoctor.convert(input, { doctype: 'inline', extension_registry: registry })
    expect(output).to.equal('<sup class="fragment highlight-red">text</sup>')
  })
  it('should apply subscript formatting on inline phrases', () => {
    const registry = asciidoctor.Extensions.create()
    phraseInlineMacroExtension.register(registry)
    const input = 'phrase:subscript[text,role="fragment highlight-red"]'
    const output = asciidoctor.convert(input, { doctype: 'inline', extension_registry: registry })
    expect(output).to.equal('<sub class="fragment highlight-red">text</sub>')
  })
  it('should apply formatting on inline phrases', () => {
    const registry = asciidoctor.Extensions.create()
    phraseInlineMacroExtension.register(registry)
    const input = 'phrase:#[text,role="fragment highlight-red"]'
    const output = asciidoctor.convert(input, { doctype: 'inline', extension_registry: registry })
    expect(output).to.equal('<span class="fragment highlight-red">text</span>')
  })
  describe('Load document', () => {
    it('should preserve attributes', () => {
      class CustomHtml5Converter {
        constructor () {
          this.baseConverter = asciidoctor.Html5Converter.$new()
        }
        convert (node, transform) {
          if (node.getNodeName() === 'inline_quoted') {
            const dataAttrs = Object.entries(node.getAttributes())
              .filter(([key, _]) => key.startsWith('data-'))
              .map(([key, value]) => `${key}="${value}"`)
              .join(' ')
            return `<${node.getType()}${dataAttrs ? ` ${dataAttrs}` : ''}>${node.getText()}</${node.getType()}>`
          }
          return this.baseConverter.convert(node, transform)
        }
      }
      asciidoctor.ConverterFactory.register(new CustomHtml5Converter(), ['custom-html5'])
      const registry = asciidoctor.Extensions.create()
      phraseInlineMacroExtension.register(registry)
      const input = 'phrase:strong[text,data-fragment-index=1]'
      const doc = asciidoctor.load(input, { doctype: 'inline', extension_registry: registry, backend: 'custom-html5' })
      expect(doc.convert()).to.equal('<strong data-fragment-index="1">text</strong>')
    })
  })
})
